src/rust/graph-query-service/src/graph_query.rs

```diff
@@ -0,0 +1,87 @@
+use futures::future::join_all;
+pub use rust_proto::graplinc::grapl::api::graph_query_service::v1beta1::messages::StringCmp;
+use rust_proto::graplinc::grapl::{
+    api::graph_query_service::v1beta1::messages::{
+        GraphQuery,
+        GraphView,
+    },
+    common::v1beta1::types::Uid,
+};
+
+use crate::{
+    node_query::{
+        fetch_node_with_edges,
+        NodeQueryError,
+    },
+    property_query::PropertyQueryExecutor,
+    short_circuit::ShortCircuit,
+    visited::Visited,
+};
+
+#[derive(thiserror::Error, Debug)]
+pub enum GraphQueryError {
+    #[error("NodeQueryError {0}")]
+    NodeQueryError(#[from] NodeQueryError),
+}
+
+#[tracing::instrument(skip(graph_query, property_query_executor))]
+pub async fn query_graph(
+    graph_query: &GraphQuery,
+    uid: Uid,
+    tenant_id: uuid::Uuid,
+    property_query_executor: PropertyQueryExecutor,
+) -> Result<Option<(GraphView, Uid)>, GraphQueryError> {
+    let mut query_handles = Vec::with_capacity(graph_query.node_property_queries.len());
+    let x_query_short_circuiter = ShortCircuit::new();
+    // We should add a way for different queries to short circuit each other
+    for node_query in graph_query.node_property_queries.values() {
```

Comment:
I *think* `values` could be `into_values` and some clones can get removed.

Reply:
Won't work for now, graph_query is a reference and we later pass it into other functions that expect a full un-partial-moved GraphQuery.